manta-crypto/src/arkworks/serialize.rs

```diff
@@ -25,6 +28,153 @@ use {
 #[doc(inline)]
 pub use ark_serialize::*;
 
+/// Serialization Hook
+pub trait HasSerialization<'s>: 's {
+    /// Serialize Type
+    type Serialize: CanonicalSerialize + From<&'s Self>;
+}
+
+/// Deserialization Hook
+pub trait HasDeserialization: Sized {
+    /// Deserialize Type
+    type Deserialize: CanonicalDeserialize + Into<Self>;
+}
+
+/// Arkworks Reader
+pub struct ArkReader<R>
+where
+    R: codec::Read,
+{
+    /// Reader State
+    state: Result<R, R::Error>,
+}
+
+impl<R> ArkReader<R>
+where
+    R: codec::Read,
+{
+    /// Builds a new [`ArkReader`] from `reader`.
+    #[inline]
+    pub fn new(reader: R) -> Self {
+        Self { state: Ok(reader) }
+    }
+
+    /// Updates the internal reader state by performing the `f` computation.
+    #[inline]
+    fn update<T, F>(&mut self, f: F) -> Option<T>
+    where
+        F: FnOnce(&mut R) -> Result<T, R::Error>,
+    {
+        if let Ok(reader) = self.state.as_mut() {
+            match f(reader) {
+                Ok(value) => return Some(value),
+                Err(err) => self.state = Err(err),
+            }
+        }
+        None
+    }
+
+    /// Returns the reader state back or an error if it occured during any [`Read`](io::Read)
+    /// methods.
```

Comment:
The reader state is already defined as the result, not `R`.

Reply:
Same as writer state.